[ git-webkit ] Add modify command

robert-jenner opened the pull request:
#### 97ea4aa6921469a9407ffd44190c0d200eda6dea
<pre>
[ git-webkit ] Add modify command
<a href="https://bugs.webkit.org/show_bug.cgi?id=292636">https://bugs.webkit.org/show_bug.cgi?id=292636</a>
<a href="https://rdar.apple.com/150801776">rdar://150801776</a>

Reviewed by NOBODY (OOPS!).

Add a command to quickly modify a radar when given a commit hash.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/__init__.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py: Added.
(add_keyword_to_issue):
(get_keyword):
(Modify):
(Modify.parser):
(Modify.main):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/modify_unittest.py: Added.
(TestModify):
(TestModify.setUp):
(TestModify.test_basic):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/97ea4aa6921469a9407ffd44190c0d200eda6dea

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135260 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7647 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46527 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142766 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/87012 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4a86e607-8fa3-4a4a-8c38-a9f745b4ed7a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137129 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8283 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7494 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103364 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/70742 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/af3e2142-1a8f-46db-a5d3-1d83094bec7e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138206 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5922 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121234 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84226 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/0e1d463f-b413-444d-b5d5-d0bd0576e0dd) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/134608 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5709 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3308 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3353 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114913 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39414 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145461 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7329 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39983 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111743 "Passed tests") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/134687 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7373 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6142 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112110 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5552 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117532 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/61264 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7383 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35661 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7138 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7359 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7241 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

briannafan commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py:
```diff
@@ -0,0 +1,127 @@
+
+        def modify_radar(
+            keyword=keyword,
+            action=lambda: None,
```
We can delete this parameter and also the function call in line 121 since we don't actually use it

robert-jenner replied:
We found that this is needed, so it wasn't removed. But the print argument was.

briannafan replied:
Oh sorry, I meant the lines `action=lambda: None` and  `action()`!

briannafan commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py:
```diff
@@ -0,0 +1,127 @@
+            keyword=keyword,
+            action=lambda: None,
+        ):
+            print(keyword)
```
We can delete this print statement since we're no longer debugging. You can also add other print statements throughout your code if you want to let the user know what's happening while the code is running! 

(For example, `print(f"Adding '{keyword}' to {issue.link}...")` will print `Adding 'WebKit Merge-back Approved' to rdar://12345`.)

briannafan commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py:
```diff
@@ -0,0 +1,127 @@
+                sys.stderr.write('Radar is not available on this machine\n')
+            return 255
+
+        # First, check if we were provided a bug URL
```
Lines 79-85 check if the first argument provided is a bug URL. However, your script is explicit about only passing in commit hashes as the first argument so we can delete these lines and unindent lines 86-105.

It doesn't actively hurt us to keep this code, but it's also redundant, so we should remove it for clarity and conciseness.

briannafan commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py:
```diff
@@ -0,0 +1,127 @@
+                rdar = tracker
+                break
+        if not rdar or not rdar.radarclient():
+            sys.stderr.write("'cloning' is a concept which belongs solely to radar\n")
```
Can remove this line since we're not cloning

briannafan commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py:
```diff
@@ -0,0 +1,127 @@
+
+    @classmethod
+    def parser(cls, parser, loggers=None):
+        parser.add_argument(
```
Following up on our discussion on how to modify this script to take in multiple commits:
- [revert.py](https://github.com/WebKit/WebKit/blob/3403fdd56fabdebf93b49797f786e328a2ee57d4/Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py#L49) is a good example of how to create an argument with multiple values. (Important part is `nargs=+`.) It would be called like `git-webkit revert 12345 12346 123458`
- Once we have the argument, we can iterate through them to access each commit value 
```
for commit_hash in args.argument:
      <your logic here>
```
- It would be helpful to think about which parts of your code need to be run per commit and that is what will run in the loop

briannafan reviewed (commented):
Looking good so far! My comments are mainly about cleaning up unnecessary code and further actions you could take to add logging and multiple commits.

briannafan commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py:
```diff
@@ -0,0 +1,102 @@
+        parser.add_argument(
+            'arguments', nargs='+',
+            type=str,
+            help='String representation of the commit hash(s) to modify',
```
nit: Plural of hash is hashes

briannafan commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py:
```diff
@@ -0,0 +1,102 @@
+        for commit_hash in args.arguments:
+            commit = repository.find(commit_hash, include_identifier=False)
+            if not commit:
+                sys.stderr.write(f"Failed to resolve '{commit_hash}' to a radar issue\n")
```
This error message is not entirely accurate. We haven't searched for the radar yet, this is just looking for a commit object given a commit hash.

Maybe `f"Failed to find '{commit_hash}' in the current repository\n"`.

briannafan commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py:
```diff
@@ -0,0 +1,102 @@
+                break
+        if not rdar or not rdar.radarclient():
+            if not rdar:
+                sys.stderr.write('Radar not available. Please validate crednetials. \n')
```
typo: `crednetials` -> `credentials`

briannafan commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py:
```diff
@@ -0,0 +1,102 @@
+                sys.stderr.write(f"Found {commit}, but it doesn't reference any issues\n")
+                continue
+
+            issue = next((c for c in commit.issues if isinstance(c.tracker, radar.Tracker)), None)
```
nit: `c` sounds like short for `commit` but we're iterating through Issue objects (that's what `commit.issues` contains). Can we rename?

briannafan commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py:
```diff
@@ -0,0 +1,102 @@
+                issue = commit.issues[0]
+
+            log.info(f"{commit} references {issue}")
+
```
You probably want to keep the code here that searches for a radar if the first issue in `commit.issues` happens to be a bugzilla.
```
if not isinstance(issue.tracker, radar.Tracker):
     for candidate in issue.references:
          if isinstance(candidate.tracker, radar.Tracker):
               issue = candidate
               break
```

briannafan commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py:
```diff
@@ -0,0 +1,102 @@
+                issue.set_keywords([keyword] + issue.keywords)
+            except radar.radarclient().exceptions.UnsuccessfulResponseException:
+                sys.stderr.write(f"Failed to add keyword to radar issue {issue}\n")
+                continue
```
This `continue` is redundant, maybe we want to replace it with `return 1` instead to show error?

briannafan commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py:
```diff
@@ -0,0 +1,102 @@
+            '--keyword', '-k',
+            dest='keyword',
+            required=True,
+            help='Keyword to add to radar from commit hash(s)',
```
Same as above

briannafan reviewed (changes requested):
Looking good! Can we add a unit test for this?

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py:
```diff
@@ -0,0 +1,108 @@
+    @classmethod
+    def main(cls, args, repository, **kwargs):
+        keyword = args.keyword
+        if not keyword:
```
Based on @robert-jenner's vision for this script, I think we want the keyword addition to exist as it's own function, since we intend to add more actions to this script in the future.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py:
```diff
@@ -0,0 +1,108 @@
+                sys.stderr.write('Radar not available. Please validate credentials. \n')
+            return 255
+
+        for commit_hash in args.arguments:
```
This looks really similar to `git-webkit find`. Does that command have something that returns a list of matching commits, maybe? Another question would be if you could make this "pipeable", ie, make something like this work:
`git log --online -20 | git-webkit modify ...`

JonWBedard replied:
As a side note, "integrate with `git-webkit find`" and "pipeable" are mutually exclusive.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py:
```diff
@@ -0,0 +1,108 @@
+                        break
+
+            try:
+                issue.set_keywords([keyword] + issue.keywords)
```
This is the bit that that should be it's own function since the "find all commits" and "find all issues for commits" are general pieces of code applicable to other modification actions.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py:
```diff
@@ -0,0 +1,119 @@
+
+def add_keyword_to_issue(issue, keyword):
+    if not isinstance(issue.tracker, radar.Tracker):
+        for candidate in issue.references:
```
Need an else here, right? Although Bugzilla also has keywords, I think, although we don't support them, if I recall.

robert-jenner replied:
I added an else statement here where if it doesn't find a radar, it exits.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py:
```diff
@@ -0,0 +1,119 @@
+    try:
+        issue.set_keywords([keyword] + issue.keywords)
+        log.info(f"Successfully added '{keyword}' to the issue {issue}")
+    except radar.radarclient().exceptions.UnsuccessfulResponseException:
```
Can you test this without `radarclient()` available? I think this would result in an exception. Need an early exist at the beginning of the function

robert-jenner replied:
Fixed this too.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py:
```diff
@@ -0,0 +1,119 @@
+
+def get_keyword(args):
+    if not args.keyword:
+        print("No keyword provided. Exiting.")
```
Should print this to standard error.

robert-jenner replied:
Fixed.

JonWBedard reviewed (commented):
A few small notes about radar/bugzilla integration.

robert-jenner committed:
[ git-webkit ] Add modify command
https://bugs.webkit.org/show_bug.cgi?id=292636
rdar://150801776

Reviewed by NOBODY (OOPS!).

Add a command to quickly modify a radar when given a commit hash.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/__init__.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py: Added.
(add_keyword_to_issue):
(get_keyword):
(Modify):
(Modify.parser):
(Modify.main):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/modify_unittest.py: Added.
(TestModify):
(TestModify.setUp):
(TestModify.test_basic):

bls1999 commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py:
```diff
@@ -0,0 +1,128 @@
+
+def get_keyword(args):
+    if not args.keyword:
+        sys.stderr.write("No keyword provided. Exiting.\n")
```
Since you have the keyword marked as required in the argument parser, this condition will never be true.

robert-jenner replied:
True. I will modify.

JonWBedard commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/modify.py:
```diff
@@ -0,0 +1,128 @@
+            return 255
+
+        for commit_hash in args.arguments:
+            commit = repository.find(commit_hash, include_identifier=False)
```
Is this intended to allow ranges too? Because it won't, although there are sub-commands that do that, if we wanted to.
